AlphaWallet/KeyManagement/EtherKeystore.swift

```diff
@@ -0,0 +1,647 @@
+    }
+
+    private func encryptionKeyForSeedLabel(fromAccount account: EthereumAccount, withUserPresence: Bool) -> String {
+        let prefix: String
+        if withUserPresence {
+            prefix = Keys.encryptionKeyForSeedUserPresenceRequiredPrefix
+        } else {
+            prefix = Keys.encryptionKeyForSeedUserPresenceNotRequiredPrefix
+        }
+        return "\(prefix)\(account.address.eip55String)"
+    }
+
+    private func encryptionKeyForPrivateKeyLabel(fromAccount account: EthereumAccount, withUserPresence: Bool) -> String {
+        let prefix: String
+        if withUserPresence {
+            prefix = Keys.encryptionKeyForPrivateKeyUserPresenceRequiredPrefix
+        } else {
+            prefix = Keys.encryptionKeyForPrivateKeyUserPresenceNotRequiredPrefix
+        }
+        return "\(prefix)\(account.address.eip55String)"
+    }
+
+    func elevateSecurity(forAccount account: EthereumAccount) -> Bool {
+        guard !isProtectedByUserPresence(account: account) else { return true }
+        guard isUserPresenceCheckPossible else { return false }
+        let prompt: String
+        var isSuccessful: Bool
+        if isHdWallet(account: account) {
+            prompt = R.string.localizable.keystoreAccessKeyHdLock()
+            guard let seed = getSeedForHdWallet(forAccount: account, prompt: prompt, withUserPresence: false) else { return false }
+            isSuccessful = saveSeedForHdWallet(seed, forAccount: account, withUserPresence: true)
+            if isSuccessful {
+                //Read it back, forcing iOS to check for user-presence
+                isSuccessful = getSeedForHdWallet(forAccount: account, prompt: prompt, withUserPresence: true) != nil
+            }
+        } else {
+            prompt = R.string.localizable.keystoreAccessKeyNonHdLock()
+            guard let keyStoredAsRawPrivateKey = getPrivateKeyFromNonHdWallet(forAccount: account, prompt: prompt, withUserPresence: false) else { return false }
+            isSuccessful = savePrivateKeyForNonHdWallet(keyStoredAsRawPrivateKey, forAccount: account, withUserPresence: true)
+            if isSuccessful {
+                //Read it back, forcing iOS to check for user-presence
+                isSuccessful = getPrivateKeyFromNonHdWallet(forAccount: account, prompt: prompt, withUserPresence: true) != nil
+            }
+        }
+        if isSuccessful {
+            addToListOfEthereumAddressesProtectedByUserPresence(account.address)
+            let secureEnclave = SecureEnclave()
+            if isHdWallet(account: account) {
+                secureEnclave.deletePrivateKeys(withName: encryptionKeyForSeedLabel(fromAccount: account, withUserPresence: false))
+                keychain.delete("\(Keys.ethereumSeedUserPresenceNotRequiredPrefix)\(account.address.eip55String)")
```

Comment:
Can secureEnclave.deletePrivateKeys throw an exception? In that case, we will never execute `keychain.delete`. Ideally, the second deletion should not be contingent on the previous one succeeding, i.e., we should delete as much as possible rather than "aborting" if a preceding operation fails.

Reply:
No, `SecureEnclave.deletePrivateKeys` will not throw an exception that can be caught (like Java). If it does, it will have to marked with the `throws` keyword in its signature, like:

```
func deletePrivateKeys(withName name: String) throws {}
```

And the compiler will force it to be handled (do-catch, like Java's try-catch) or explicitly ignored with a `try?` like:

```
let result = try? secureEnclave.deletePrivateKeys(...) //ignores error if there's one, `result` will be nil
```

Or power through and crash upon error with `try!`:

```
let result = try! secureEnclave.deletePrivateKeys(...) //crash if error is thrown
```

Or propogate it upstream with `try`:

```
let result = try secureEnclave.deletePrivateKeys(...) //have to handle it with do-catch locally or mark function with `throws`
```

It is also possible that a Swift function generates a runtime error (corruption of the stack for example), but those typically result in a crash and we can't recover from them programmatically other than grabbing a crash log.

So Swift is more of a very safe C, but not as safe as VM-based languages like Java in this aspect.

Reply:
Again, thanks for clarifying.